lpm/src/prefix/with_ports.rs

```diff
@@ -0,0 +1,299 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use crate::prefix::{Prefix, PrefixSize};
+use bnum::BUint;
+use std::fmt::Display;
+
+/// A type for the size of IP/port combinations for IP prefixes with associated port ranges.
+/// We need something larger than u128 to cover all possible combinations: the maximum value is
+/// `(u128::MAX + 1) * (u16::MAX + 1)`.
+pub type PrefixWithPortsSize = BUint<3>;
+
+/// Trait for IP ranges (CIDR prefix or simple ranges) with associated port ranges.
+pub trait IpRangeWithPorts {
+    /// Returns the address range length.
+    fn addr_range_length(&self) -> PrefixSize;
+    /// Returns the port range length.
+    fn port_range_length(&self) -> usize;
+    /// Returns true if the range fully covers (contains) the given other range.
+    fn covers(&self, other: &Self) -> bool;
+    /// Returns true if the range overlaps with the given other range.
+    fn overlaps(&self, other: &Self) -> bool;
+    /// Returns the total number of (IP, port) combinations covered by the IP and port ranges.
+    ///
+    /// # Warning
+    ///
+    /// If the port range is empty, this method should return the number of addresses in the address
+    /// range: so this will produce a different value than an object with the same IP range, but
+    /// with a port range covering all existing ports.
+    fn size(&self) -> PrefixWithPortsSize {
```

Comment:
Why make this semantic distinction?  is it just to allow compatibility between an ip range and an ip range with ports?

In other words, if this trait is called IpRangeWithPorts, then should it always count (Ip, port) pairs available, so no port ranges means the size is the same as all port ranges?

If we what this to mean something like IpRangeOrIpRangeWithPorts (terrible name, we should choose something shorter), then the better way to say this is that it returns the available size of items.  For an IpRange that would be the number of IPs for an IPPortRange, that would be a different thing entirely.

Perhaps a discussion on the correct design here is in order.

Reply:
The comment is in fact outdated. I had a previous version of the code that made a distinction, but it's been changed a while ago. I'll remove the comment.

Reply:
Comment removed